include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,665 @@
+{
+  additional_data_nox = additional_data;
+  parameters_nox      = parameters;
+}
+#endif
+
+#ifdef DEAL_II_WITH_SUNDIALS
+template <typename VectorType>
+void
+NonlinearSolverSelector<VectorType>::set_data(
+  const typename SUNDIALS::KINSOL<VectorType>::AdditionalData &additional_data)
+{
+  additional_data_kinsol = additional_data;
+}
+#endif
+
+
+template <typename VectorType>
+void
+NonlinearSolverSelector<VectorType>::solve_with_petsc(
+  VectorType & /*initial_guess_and_solution*/)
+{
+  AssertThrow(false,
+              ExcMessage("PETSc SNES requires you to use PETSc vectors."));
+}
+
+
+
+#ifdef DEAL_II_WITH_PETSC
+template <>
+void
+NonlinearSolverSelector<PETScWrappers::MPI::Vector>::solve_with_petsc(
+  PETScWrappers::MPI::Vector &initial_guess_and_solution)
+{
+  PETScWrappers::NonlinearSolver<PETScWrappers::MPI::Vector> nonlinear_solver(
+    additional_data_petsc_snes, mpi_communicator);
+
+  nonlinear_solver.residual = residual;
+
+  nonlinear_solver.setup_jacobian = setup_jacobian;
+
+  nonlinear_solver.solve_with_jacobian =
+    [&](const PETScWrappers::MPI::Vector &src,
+        PETScWrappers::MPI::Vector &      dst) -> int {
+    const double tolerance = 1e-5;
+    return solve_with_jacobian(src, dst, tolerance);
+  };
+
+  nonlinear_solver.solve(initial_guess_and_solution);
+}
+#endif
+
+
+
+template <typename VectorType>
+void
+NonlinearSolverSelector<VectorType>::solve(
+  VectorType &initial_guess_and_solution)
+{
+  // The "auto" solver_type will default to kinsol, however if KINSOL is not
+  // available then we will use NOX.
+  if (additional_data.solver_type == AdditionalData::SolverType::automatic)
+    {
+#ifdef DEAL_II_WITH_PETSC
+      additional_data.solver_type = AdditionalData::SolverType::petsc_snes;
+#endif
+#ifdef DEAL_II_WITH_TRILINOS
+      additional_data.solver_type = AdditionalData::SolverType::nox;
+#endif
+#ifdef DEAL_II_WITH_SUNDIALS
+      additional_data.solver_type = AdditionalData::SolverType::kinsol;
+#endif
+
+      // If "auto" is still the solver type we cannot solve the problem
+      if (additional_data.solver_type == AdditionalData::SolverType::automatic)
+        AssertThrow(false, ExcMessage("No valid solver type."));
+    }
+
+  if (additional_data.solver_type == AdditionalData::SolverType::kinsol)
+    {
+#ifdef DEAL_II_WITH_SUNDIALS
+      SUNDIALS::KINSOL<VectorType> nonlinear_solver(additional_data_kinsol,
+                                                    mpi_communicator);
+
+      // We set the KINSOL reinit vector equal to the same function
+      // defined for NonlinearSolverSelector.
+      nonlinear_solver.reinit_vector = reinit_vector;
+
+      nonlinear_solver.residual = residual;
+
+      // We cannot simply set these two functions equal to each other
+      // because they have a different number of inputs.
+      nonlinear_solver.setup_jacobian = [&](const VectorType &current_u,
+                                            const VectorType /*&current_f*/) {
```

Comment:
This would have cost you a copy of a vector :-)
```suggestion
                                            const VectorType & /*current_f*/) {
```